sympy/core/operations.py

```diff
@@ -504,3 +505,118 @@ def args(self):
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     if issubclass(_1, NewExpr):
+    ...         return _1
+    ...     else:
+    ...         return _2
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+
+    See Also
+    ========
+
+    core.compatibility.OpDispatcher
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.name = name
+        self.fallback = fallback
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self.dispatcher = Dispatcher(name)
+
+    def __repr__(self):
+        return "<dispatched %s>" % self.name
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the type which has the handler that is to be used.
+        """
+        def _(func):
+            self.add_priority((cls1, cls2), func)
+            return func
+        return _
+
+    def add_priority(self, signature, func, on_ambiguity=(lambda *args, **kwargs: None)):
+        # This suppresses unnecessary ambiguity warning, since not all sympy objects
+        # need to operate with each other.
+        return self.dispatcher.add(signature, func, on_ambiguity)
+
+    def __call__(self, *args, **kwargs):
+        types = tuple([type(arg) for arg in args])
+        return self.dispatch(types)(*args, **kwargs)
+
+    def dispatch(self, types):
+        """
+        Select the argument with highest priority, and return its handler.
+        """
+        args_data = []
+        for typ in types:
+            handler = getattr(typ, self.handlermethod, None)
+            if handler is None:
+                continue
+            op_priority = getattr(typ, '_op_priority', 0)
+            args_data.append((typ, op_priority, handler))
+
+        # No handler detected: use fallback
+        if len(args_data) == 0:
+            return self.fallback
+
+        # Sieve with _op_priority first
+        max_priority = max(op_priority for _, op_priority, _ in args_data)
+        sieved_data = [tup for tup in args_data if tup[1]==max_priority]
+
+        # Recursively select with registered binary priority
+        for i, tup in enumerate(sieved_data):
+            if i == 0:
+                result = tup
+            else:
+                selected = self.dispatcher.dispatch(result[0], tup[0])(result[0], tup[0])
+                if selected is tup[0]:
+                    result = tup
+                elif selected is result[0]:
+                    pass
+                else:
+                    # Cannot determine priority
+                    raise RuntimeError("Dispatcher {!r} and {!r} returned invalid result {!r}.".format(result[0], tup[0], selected))
```

Comment:
To support my suggestion below:
```suggestion
                if not isinstance(selected, type):
                    raise RuntimeError("Dispatcher for {!r} and {!r} must return a type, got {!r}".format(result[0], tup[0], selected)
                handler = getattr(selected, self.handlermethod, None)
                if handler is None:
                    raise RuntimeError("Dispatcher for {!r} and {!r} returned a type with no handler, {!r}.".format(result[0], tup[0], selected))
                result = (selected, None, handler)
```